File: onebusaway-android/src/main/java/org/onebusaway/android/ui/widget/StopTimesWidgetConfigActivity.java
/*
 * Copyright (C) 2025 Rob Godfrey (rob_godfrey@outlook.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.ui.widget;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import org.onebusaway.android.R;
import org.onebusaway.android.io.elements.ObaArrivalInfo;
import org.onebusaway.android.io.request.ObaArrivalInfoRequest;
import org.onebusaway.android.io.request.ObaArrivalInfoResponse;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.graphics.Rect;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.view.inputmethod.InputMethodManager;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.widget.EditText;
import android.widget.RemoteViews;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

public class StopTimesWidgetConfigActivity extends AppCompatActivity {

    private static class Route {
        final String id;
        final String shortName;

        Route(String id, String shortName) {
            this.id = id;
            this.shortName = shortName;
        }
    }

    public static final String EXTRA_STOP_ID = "stop_id";
    public static final String EXTRA_STOP_NAME = "stop_name";
    private static final int MAX_MINUTES_AFTER = 1440;
    private int appWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID;

    private EditText widgetNameInput;
    private MaterialButton selectStopButton;
    private MaterialButton saveButton;
    private TextView routesLoadingText;
    private ChipGroup chipGroup;

    private String selectedStopId = null;
    private String selectedStopName = null;
    private String existingWidgetName = null;
    private Set<String> existingRouteFilter = null;

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final AtomicInteger mLoadGeneration = new AtomicInteger();

    private final ActivityResultLauncher<Intent> mStopPickerLauncher =
            registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
                if (result.getResultCode() == RESULT_OK && result.getData() != null) {
                    selectedStopId = result.getData().getStringExtra(StopPickerActivity.EXTRA_STOP_ID);
                    selectedStopName = result.getData().getStringExtra(StopPickerActivity.EXTRA_STOP_NAME);
                    updateStopButton();
                    widgetNameInput.setText(selectedStopName);
                    updateSaveButton();
                    loadRoutesForStop(selectedStopId);
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        supportRequestWindowFeature(Window.FEATURE_NO_TITLE);

        setResult(RESULT_CANCELED); // widget is not added unless user hits the "save" button
        setContentView(R.layout.stop_times_widget_config);

        appWidgetId = getIntent().getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        selectedStopId = getIntent().getStringExtra(EXTRA_STOP_ID);
        selectedStopName = getIntent().getStringExtra(EXTRA_STOP_NAME);

        // If editing an existing widget, load the saved config to populate fields
        if (selectedStopId == null && appWidgetId != AppWidgetManager.INVALID_APPWIDGET_ID) {
            final WidgetConfig existing = WidgetPrefs.loadConfig(this, appWidgetId);
            if (existing != null) {
                selectedStopId = existing.getStopId();
                selectedStopName = existing.getStopName();
                existingWidgetName = existing.getWidgetName();
                existingRouteFilter = existing.getRoutes();
            }
        }

        setupUi();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        executor.shutdownNow();
    }

    // Hide keyboard when user taps screen if focus is on the 'Widget name' EditText
    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            final View focused = getCurrentFocus();
            if (focused instanceof EditText) {
                final Rect rect = new Rect();
                focused.getGlobalVisibleRect(rect);
                if (!rect.contains((int) event.getRawX(), (int) event.getRawY())) {
                    focused.clearFocus();
                    ((InputMethodManager) getSystemService(INPUT_METHOD_SERVICE))
                            .hideSoftInputFromWindow(focused.getWindowToken(), 0);
                }
            }
        }
        return super.dispatchTouchEvent(event);
    }

    private void setupUi() {
        widgetNameInput = findViewById(R.id.widget_name_input);
        selectStopButton = findViewById(R.id.select_stop);
        saveButton = findViewById(R.id.save_button);
        routesLoadingText = findViewById(R.id.routes_loading);
        chipGroup = findViewById(R.id.route_chips);

        if (selectedStopName != null) {
            updateStopButton();
            widgetNameInput.setText(existingWidgetName != null ? existingWidgetName : selectedStopName);
        }

        widgetNameInput.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus) widgetNameInput.post(widgetNameInput::selectAll);
        });

        selectStopButton.setOnClickListener(v ->
                mStopPickerLauncher.launch(new Intent(this, StopPickerActivity.class)));

        if (selectedStopId != null) {
            loadRoutesForStop(selectedStopId);
        }

        updateSaveButton();
        saveButton.setOnClickListener(v -> onSave());
    }

    private void updateStopButton() {
        selectStopButton.setText(selectedStopName);
        selectStopButton.setIconResource(R.drawable.ic_close);
    }

    private void updateSaveButton() {
        saveButton.setEnabled(selectedStopId != null);
    }

    private void loadRoutesForStop(String stopId) {
        chipGroup.removeAllViews();
        routesLoadingText.setVisibility(View.VISIBLE);

        // Increment the generation so if the user picks a different stop before the current request
        // finishes, the stale result is discarded instead of populating the chips
        final int generation = mLoadGeneration.incrementAndGet();

        executor.execute(() -> {
            final ObaArrivalInfoResponse response =
                    ObaArrivalInfoRequest.newRequest(this, stopId, MAX_MINUTES_AFTER).call();

            final List<Route> routes = new ArrayList<>();
            final Set<String> seen = new HashSet<>();
            if (response != null && response.getArrivalInfo() != null) {
                for (ObaArrivalInfo info : response.getArrivalInfo()) {
                    if (seen.add(info.getRouteId())) {
                        routes.add(new Route(info.getRouteId(), info.getShortName()));
                    }
                }
            }
            Collections.sort(routes, (a, b) -> compareRouteNames(a.shortName, b.shortName));

            if (mLoadGeneration.get() == generation) {
                mainHandler.post(() -> populateRouteChips(routes));
            }
        });
    }

    // Sorts route names numerically if both are integers, otherwise alphabetically
    private static int compareRouteNames(String a, String b) {
        try {
            return Integer.compare(Integer.parseInt(a), Integer.parseInt(b));
        } catch (NumberFormatException e) {
            return a.compareToIgnoreCase(b);
        }
    }

    private void populateRouteChips(List<Route> routes) {
        final int green = ContextCompat.getColor(this, R.color.theme_primary);
        final int grey = ContextCompat.getColor(this, R.color.theme_muted);
        final int[][] states = {new int[]{android.R.attr.state_checked}, new int[]{-android.R.attr.state_checked}};
        final ColorStateList bgColors = new ColorStateList(states, new int[]{green, grey});
        final ColorStateList textColors = new ColorStateList(states, new int[]{Color.WHITE, Color.WHITE});

        // transparent icon to prevent chips from jumping around when the check mark appears/disappears
        final float iconSizePx = 14 * getResources().getDisplayMetrics().density;
        final android.graphics.drawable.GradientDrawable transparentIcon = new android.graphics.drawable.GradientDrawable();
        transparentIcon.setColor(Color.TRANSPARENT);
        transparentIcon.setSize((int) iconSizePx, (int) iconSizePx);

        routesLoadingText.setVisibility(View.GONE);
        chipGroup.removeAllViews();

        for (final Route route : routes) {
            final Chip chip = new Chip(this);
            chip.setText(route.shortName);
            chip.setTag(route.id);
            chip.setCheckable(true);
            chip.setChipIcon(transparentIcon);
            chip.setChipIconVisible(true);
            chip.setChipIconSize(iconSizePx);
            chip.setCheckedIconVisible(true);
            chip.setCheckedIcon(ContextCompat.getDrawable(this, R.drawable.ic_chip_check));
            chip.setChipStrokeWidth(0);
            chip.setRippleColor(ColorStateList.valueOf(Color.TRANSPARENT)); // hide ripple effect
            chip.setChipBackgroundColor(bgColors);
            chip.setTextColor(textColors);
            chip.setChipEndPadding(chip.getChipStartPadding() + iconSizePx);
            chip.ensureAccessibleTouchTarget(0);
            // Restore previous selection if editing widget, otherwise default to none selected
            boolean checked = existingRouteFilter != null && existingRouteFilter.contains(route.id);
            chip.setChecked(checked);
            chipGroup.addView(chip);
        }
    }

    private void onSave() {
        if (selectedStopId == null) return;

        final Map<String, String> routeShortNames = getSelectedRouteShortNames();
        if (routeShortNames.isEmpty()) {
            Toast.makeText(this, R.string.widget_config_select_at_least_one_route, Toast.LENGTH_SHORT).show();
            return;
        }
        if (routeShortNames.size() > 3) {
            Toast.makeText(this, R.string.widget_config_select_no_more_than_three_routes, Toast.LENGTH_SHORT).show();
            return;
        }

        final String enteredName = widgetNameInput.getText().toString().trim();
        final String widgetName = enteredName.isEmpty() ? selectedStopName : enteredName;

        if (appWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            // Launched from within the app before pinning, request pin with config in callback
            pinWidgetWithConfig(widgetName, routeShortNames);
            return;
        }

        final WidgetConfig config = new WidgetConfig(selectedStopId, selectedStopName, widgetName, routeShortNames);
        final AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(this);

        WidgetPrefs.saveConfig(this, appWidgetId, config);
        StopTimesWidget.refreshWidget(this, appWidgetManager, appWidgetId);

        final Intent result = new Intent();
        result.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        setResult(RESULT_OK, result);
        finish();
    }

    private Map<String, String> getSelectedRouteShortNames() {
        final Map<String, String> selected = new HashMap<>();
        for (int i = 0; i < chipGroup.getChildCount(); i++) {
            final Chip chip = (Chip) chipGroup.getChildAt(i);
            if (chip.isChecked()) {
                selected.put((String) chip.getTag(), chip.getText().toString());
            }
        }
        return selected;
    }

    private void pinWidgetWithConfig(String widgetName, Map<String, String> routeShortNames) {
        final AppWidgetManager mgr = AppWidgetManager.getInstance(this);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O || !mgr.isRequestPinAppWidgetSupported()) {
            Toast.makeText(this, R.string.widget_config_add_manually, Toast.LENGTH_SHORT).show();
            finish();
            return;
        }

        // Save config before requesting pin. On some Android versions, the system launches the
        // configure activity after placement, which we don't want. This lets the launch detect
        // and silently apply the already-configured settings instead of showing the config screen
        WidgetPrefs.savePendingPinConfig(this, new WidgetConfig(selectedStopId, selectedStopName, widgetName, routeShortNames));

        final Intent callbackIntent = new Intent(this, StopTimesWidget.class);
        callbackIntent.setAction(StopTimesWidget.ACTION_APPLY_PENDING_CONFIG);
        callbackIntent.putExtra(StopTimesWidget.EXTRA_STOP_ID, selectedStopId);
        callbackIntent.putExtra(StopTimesWidget.EXTRA_STOP_NAME, selectedStopName);
        callbackIntent.putExtra(StopTimesWidget.EXTRA_WIDGET_NAME, widgetName);
        callbackIntent.putStringArrayListExtra(StopTimesWidget.EXTRA_ROUTE_IDS, new ArrayList<>(routeShortNames.keySet()));
        callbackIntent.putStringArrayListExtra(StopTimesWidget.EXTRA_ROUTE_NAMES, new ArrayList<>(routeShortNames.values()));

        int flags = PendingIntent.FLAG_UPDATE_CURRENT;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            flags |= PendingIntent.FLAG_MUTABLE; // required so system can inject EXTRA_APPWIDGET_ID
        }

        final PendingIntent callbackPi = PendingIntent.getBroadcast(this, 0, callbackIntent, flags);

        Bundle previewExtras = null;
        // add stop name & route ID to widget preview
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            RemoteViews preview = new RemoteViews(getPackageName(), R.layout.stop_times_widget_preview);
            preview.setTextViewText(R.id.preview_stop_name, widgetName);
            for (int i = 0; i < chipGroup.getChildCount(); i++) {
                final Chip chip = (Chip) chipGroup.getChildAt(i);
                if (chip.isChecked()) {
                    preview.setTextViewText(R.id.widget_route_1_title, chip.getText());
                    break;
                }
            }
            previewExtras = new Bundle();
            previewExtras.putParcelable(AppWidgetManager.EXTRA_APPWIDGET_PREVIEW, preview);
        }

        mgr.requestPinAppWidget(new ComponentName(this, StopTimesWidget.class), previewExtras, callbackPi);
        finish();
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/widget/StopTimesWidgetConfigActivity.kt
/*
 * Copyright (C) 2026 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.ui.widget

import android.app.Activity
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.widget.RemoteViews
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.compose.setContent
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import kotlin.random.Random
import org.onebusaway.android.R
import org.onebusaway.android.api.data.StopArrivalsDataSource
import org.onebusaway.android.ui.arrivals.DefaultArrivalsRepository
import org.onebusaway.android.ui.compose.components.ObaTopAppBar
import org.onebusaway.android.ui.compose.theme.ObaTheme
import org.onebusaway.android.ui.icons.AppIcons

/** A route the config screen offers as a chip: [id] is the route id, [shortName] the display label. */
private data class RouteOption(val id: String, val shortName: String)

private const val MAX_SELECTED_ROUTES = 3

/**
 * Configures a single Stop Times widget instance: pick a stop (via [StopPickerActivity]), pick up to
 * [MAX_SELECTED_ROUTES] of its routes, and name the widget. Saving either updates an existing
 * instance's [WidgetConfig] (reconfigure flow) or, when launched pre-placement with no `appWidgetId`
 * yet, requests the system pin a new one ([pinWidgetWithConfig]).
 */
@AndroidEntryPoint
class StopTimesWidgetConfigActivity : ComponentActivity() {

    @Inject
    lateinit var stopArrivalsDataSource: StopArrivalsDataSource

    private var appWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setResult(RESULT_CANCELED) // The widget is not added unless the user hits Save.

        appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)

        var stopId = intent.getStringExtra(EXTRA_STOP_ID)
        var stopName = intent.getStringExtra(EXTRA_STOP_NAME)
        var existingWidgetName: String? = null
        var existingRoutes: Map<String, String> = emptyMap()

        // If editing an existing widget (opened via its "reconfigure" affordance, with no stop extras
        // of its own), load the saved config to pre-populate the fields.
        if (stopId == null && appWidgetId != AppWidgetManager.INVALID_APPWIDGET_ID) {
            WidgetPrefs.loadConfig(this, appWidgetId)?.let { existing ->
                stopId = existing.stopId
                stopName = existing.stopName
                existingWidgetName = existing.widgetName
                existingRoutes = existing.routeShortNames
            }
        }

        setContent {
            ObaTheme {
                StopTimesWidgetConfigScreen(
                    initialStopId = stopId,
                    initialStopName = stopName,
                    initialWidgetName = existingWidgetName ?: stopName.orEmpty(),
                    initialSelectedRoutes = existingRoutes,
                    loadRoutes = ::loadRoutesForStop,
                    onSave = ::onSave,
                    onCancel = ::finish
                )
            }
        }
    }

    // Built from the stop's own route metadata + the response's reference pool, not the arrivals list
    // itself — a stop with no arrival in the fetch window (off-hours for a weekend/holiday-only route,
    // or a stale stop whose service was later restructured) would otherwise offer no routes to pick.
    private suspend fun loadRoutesForStop(stopId: String): List<RouteOption> {
        val response = stopArrivalsDataSource.arrivals(stopId, DefaultArrivalsRepository.MINUTES_AFTER_MAX).getOrNull()
        val routeIds = response?.stop?.routeIds.orEmpty()
        return routeIds
            .mapNotNull { id -> response?.route(id)?.let { RouteOption(id, it.shortName ?: id) } }
            .sortedWith { a, b -> compareRouteNames(a.shortName, b.shortName) }
    }

    private fun onSave(stopId: String, stopName: String, widgetName: String, selectedRoutes: Map<String, String>) {
        if (appWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            pinWidgetWithConfig(stopId, stopName, widgetName, selectedRoutes)
            return
        }

        WidgetPrefs.saveConfig(this, appWidgetId, WidgetConfig(stopId, stopName, widgetName, selectedRoutes))
        StopTimesWidget.refreshWidget(this, AppWidgetManager.getInstance(this), appWidgetId)

        setResult(RESULT_OK, Intent().putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId))
        finish()
    }

    private fun pinWidgetWithConfig(stopId: String, stopName: String, widgetName: String, selectedRoutes: Map<String, String>) {
        val appWidgetManager = AppWidgetManager.getInstance(this)
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O || !appWidgetManager.isRequestPinAppWidgetSupported) {
            Toast.makeText(this, R.string.widget_config_add_manually, Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        // Save the config before requesting the pin. On some Android versions the system launches the
        // configure activity after placement anyway, which we don't want; StopTimesWidget.onUpdate
        // detects and silently applies this instead of showing the config screen again.
        WidgetPrefs.savePendingPinConfig(this, WidgetConfig(stopId, stopName, widgetName, selectedRoutes))

        val callbackIntent = Intent(this, StopTimesWidget::class.java).apply {
            action = StopTimesWidget.ACTION_APPLY_PENDING_CONFIG
            putExtra(StopTimesWidget.EXTRA_STOP_ID, stopId)
            putExtra(StopTimesWidget.EXTRA_STOP_NAME, stopName)
            putExtra(StopTimesWidget.EXTRA_WIDGET_NAME, widgetName)
            putStringArrayListExtra(StopTimesWidget.EXTRA_ROUTE_IDS, ArrayList(selectedRoutes.keys))
            putStringArrayListExtra(StopTimesWidget.EXTRA_ROUTE_NAMES, ArrayList(selectedRoutes.values))
        }

        var flags = PendingIntent.FLAG_UPDATE_CURRENT
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            flags = flags or PendingIntent.FLAG_MUTABLE // required so the system can inject EXTRA_APPWIDGET_ID
        }
        // A random (not fixed) request code: two concurrent pin flows (e.g. the user backs out of one and
        // starts another before the first is placed) must not resolve to the same underlying
        // PendingIntent, which — since FLAG_UPDATE_CURRENT keeps the object but replaces its extras —
        // would let the second flow's config silently overwrite the first's callback before it fires.
        val requestCode = Random.nextInt()
        val callbackPendingIntent = PendingIntent.getBroadcast(this, requestCode, callbackIntent, flags)

        val previewExtras = Bundle().apply {
            val preview = RemoteViews(packageName, R.layout.stop_times_widget_preview)
            preview.setTextViewText(R.id.preview_stop_name, widgetName)
            selectedRoutes.values.firstOrNull()?.let { preview.setTextViewText(R.id.widget_route_1_title, it) }
            putParcelable(AppWidgetManager.EXTRA_APPWIDGET_PREVIEW, preview)
        }

        appWidgetManager.requestPinAppWidget(ComponentName(this, StopTimesWidget::class.java), previewExtras, callbackPendingIntent)
        finish()
    }

    companion object {
        const val EXTRA_STOP_ID = "stop_id"
        const val EXTRA_STOP_NAME = "stop_name"
    }
}

/** Sorts route names numerically if both are integers, otherwise alphabetically (case-insensitive). */
private fun compareRouteNames(a: String, b: String): Int {
    val aInt = a.toIntOrNull()
    val bInt = b.toIntOrNull()
    return if (aInt != null && bInt != null) aInt.compareTo(bInt) else a.compareTo(b, ignoreCase = true)
}

@Composable
private fun StopTimesWidgetConfigScreen(
    initialStopId: String?,
    initialStopName: String?,
    initialWidgetName: String,
    initialSelectedRoutes: Map<String, String>,
    loadRoutes: suspend (String) -> List<RouteOption>,
    onSave: (stopId: String, stopName: String, widgetName: String, selectedRoutes: Map<String, String>) -> Unit,
    onCancel: () -> Unit
) {
    val context = LocalContext.current

    var stopId by remember { mutableStateOf(initialStopId) }
    var stopName by remember { mutableStateOf(initialStopName) }
    var widgetName by remember { mutableStateOf(initialWidgetName) }
    var routes by remember { mutableStateOf<List<RouteOption>>(emptyList()) }
    var isLoadingRoutes by remember { mutableStateOf(false) }
    var selectedRouteIds by remember { mutableStateOf(initialSelectedRoutes.keys) }
    // The previously-saved id -> short name map for the stop being reconfigured. A route that has no
    // upcoming arrival in the fetch below (last bus of the night, weekend-only route, ...) won't appear
    // in `routes` at all, so it needs this fallback to survive an unrelated save instead of silently
    // dropping out of the widget's config. Cleared whenever a *different* stop is picked, since these
    // ids belong to the old stop's routes.
    var persistedRoutes by remember { mutableStateOf(initialSelectedRoutes) }

    val pickerLauncher = rememberLauncherForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode != Activity.RESULT_OK) return@rememberLauncherForActivityResult
        val data = result.data ?: return@rememberLauncherForActivityResult
        val pickedId = data.getStringExtra(StopPickerActivity.EXTRA_STOP_ID) ?: return@rememberLauncherForActivityResult
        val pickedName = data.getStringExtra(StopPickerActivity.EXTRA_STOP_NAME).orEmpty()
        stopId = pickedId
        stopName = pickedName
        widgetName = pickedName
        selectedRouteIds = emptySet()
        persistedRoutes = emptyMap()
    }

    LaunchedEffect(stopId) {
        val id = stopId ?: return@LaunchedEffect
        isLoadingRoutes = true
        routes = loadRoutes(id)
        isLoadingRoutes = false
    }

    Scaffold(topBar = { ObaTopAppBar(title = stringResource(R.string.widget_config_title), onBack = onCancel) }) { padding ->
        Column(
            modifier = Modifier.fillMaxSize().padding(padding).padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Text(stringResource(R.string.widget_config_selected_stop), style = MaterialTheme.typography.titleSmall)

            OutlinedButton(
                onClick = { pickerLauncher.launch(Intent(context, StopPickerActivity::class.java)) },
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(stopName ?: stringResource(R.string.widget_config_select_a_stop))
            }

            OutlinedTextField(
                value = widgetName,
                onValueChange = { widgetName = it },
                label = { Text(stringResource(R.string.widget_config_widget_name)) },
                singleLine = true,
                keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Sentences),
                modifier = Modifier.fillMaxWidth()
            )

            Text(stringResource(R.string.widget_config_selected_routes), style = MaterialTheme.typography.titleSmall)

            if (isLoadingRoutes) {
                CircularProgressIndicator()
            } else {
                FlowRow(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    routes.forEach { route ->
                        val checked = route.id in selectedRouteIds
                        FilterChip(
                            selected = checked,
                            onClick = {
                                selectedRouteIds = when {
                                    checked -> selectedRouteIds - route.id
                                    selectedRouteIds.size >= MAX_SELECTED_ROUTES -> {
                                        Toast.makeText(context, R.string.widget_config_select_no_more_than_three_routes, Toast.LENGTH_SHORT).show()
                                        selectedRouteIds
                                    }
                                    else -> selectedRouteIds + route.id
                                }
                            },
                            label = { Text(route.shortName) },
                            leadingIcon = if (checked) {
                                { Icon(AppIcons.Check, contentDescription = null) }
                            } else {
                                null
                            }
                        )
                    }
                }
            }

            Spacer(Modifier.weight(1f))

            Button(
                onClick = {
                    val finalStopId = stopId
                    val finalStopName = stopName
                    if (finalStopId == null || finalStopName == null) return@Button
                    if (selectedRouteIds.isEmpty()) {
                        Toast.makeText(context, R.string.widget_config_select_at_least_one_route, Toast.LENGTH_SHORT).show()
                        return@Button
                    }
                    val finalWidgetName = widgetName.trim().ifEmpty { finalStopName }
                    // Prefer the freshly-fetched short name (keeps it current if it changed); fall back
                    // to the persisted one for a selected route this fetch didn't happen to return.
                    val fetchedById = routes.associateBy { it.id }
                    val selectedRoutes = selectedRouteIds.associateWith { id ->
                        fetchedById[id]?.shortName ?: persistedRoutes[id] ?: id
                    }
                    onSave(finalStopId, finalStopName, finalWidgetName, selectedRoutes)
                },
                enabled = stopId != null,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(stringResource(R.string.widget_config_save_widget))
            }
        }
    }
}
